### PR TITLE
refactor(repository/cache): add support for adding various cache clients

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Use strategy pattern to encapsulate the algorithms for loading and storing cache (based on the client type - currently only `local`).
Use factory pattern to create cache clients based on user configuration,

All file system related handling was extracted into the `local` cache client, while all other logic related to parsing/creating data remains in one central place - the cache implementation class. 

This change allows for new cache clients to be implemented without other modifications.  

Added string based config option to define the cache type (repositoryCacheType).
Added an interface each cache client should implement.

## Context
This is a pre step for implementing S3 repository cache client.
 - https://github.com/renovatebot/renovate/issues/17112

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/cache/repository/impl/client/local/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import upath from 'upath';
+import { GlobalConfig } from '../../../../../../config/global';
+import { logger } from '../../../../../../logger';
+import { outputCacheFile, readCacheFile } from '../../../../../fs';
+import type { CacheClient, RepoCacheRecord } from '../../../types';
+
+export class LocalRepoCache implements CacheClient {
+  constructor(private platform: string, private repository: string) {}
+
```
Comment (Gabriel-Ladzaretti) [inline lib/util/cache/repository/impl/client/local/index.ts:9]: This local cache client contains the extracted fs handling logic from `lib/util/cache/repository/impl/local.ts` now renamed 
`repository/impl/repository-cache-impl.ts`

File: lib/util/cache/repository/impl/repository-cache-impl.ts [hunk 2]
```diff
@@ -1,43 +1,36 @@
...
+import type { CacheClient } from '../types';
 import { RepoCacheBase } from './base';
+import { CacheClientFactory } from './client/cache-client-factory';
 
 const compress = promisify(zlib.brotliCompress);
 const decompress = promisify(zlib.brotliDecompress);
 
-export class LocalRepoCache extends RepoCacheBase {
+export class RepositoryCacheImpl extends RepoCacheBase {
+  private cacheClient: CacheClient;
```
Comment (Gabriel-Ladzaretti) [inline lib/util/cache/repository/impl/repository-cache-impl.ts:20]: This cache implementation class now contains only the logic related to handling the cached data. i.e. encoding/decoding.
saving/writing is done by the cache client.

**Review by Gabriel-Ladzaretti — COMMENTED** [review]
Largest changes listed below

File: lib/util/cache/repository/impl/client/cache-client-factory.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    switch (cacheType) {
+      case 'local':
+        this.client = new LocalRepoCache(platform!, repository);
+        break;
+      // istanbul ignore next: untestable
+      default:
+        this.client = new LocalRepoCache(platform!, repository);
+        logger.info(
+          `Repository cache type not supported using type "local" instead`
+        );
```
Comment (rarkins) [inline lib/util/cache/repository/impl/client/cache-client-factory.ts:26]:
```suggestion
        logger.warn(
          { cacheType },
          `Repository cache type not supported using type "local" instead`
        );
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Have you also checked how much changes to the app this will need- hopefully one line?

File: lib/util/cache/repository/impl/client/cache-client-factory.ts (hunk 3 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/cache-client-factory.ts:26]: i dont think any change will be required. i run locally without any changes and repo cache works like it should. 
repoCacheTypes defaults to 'local'

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -626,6 +626,8 @@ Set this to `"enabled"` to have Renovate maintain a JSON file cache per-reposito
 Set to `"reset"` if you ever need to bypass the cache and have it overwritten.
 JSON files will be stored inside the `cacheDir` beside the existing file-based package cache.
 
+## repositoryCacheType
```
Comment (nabeelsaabna) [inline docs/usage/self-hosted-configuration.md:629]: missing description ?

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -229,6 +229,16 @@ const options: RenovateOptions[] = [
     default: 'disabled',
     experimental: true,
   },
+  {
+    name: 'repositoryCacheType',
+    description:
+      'Set the type of renovate repository cache if repositoryCache is enabled.',
```
Comment (nabeelsaabna) [inline lib/config/options/index.ts:235]: repositoryCache allowedValues: disabled, enabled, reset
```suggestion
      'Set the type of renovate repository cache if repositoryCache is not disabled.',
```

File: lib/util/cache/repository/impl/client/cache-client-factory.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        break;
+    }
+
+    return this.client;
+  }
+
+  static reset(): void {
+    this.client = null;
+  }
```
Comment (nabeelsaabna) [inline lib/util/cache/repository/impl/client/cache-client-factory.ts:35]: why is this needed for the client ?

File: lib/util/cache/repository/impl/client/local/index.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  public getCacheFileName(): string {
+    const cacheDir = GlobalConfig.get('cacheDir');
+    const repoCachePath = '/renovate/repository/';
+    const platform = this.platform;
+    const fileName = `${this.repository}.json`;
+    return upath.join(cacheDir, repoCachePath, platform, fileName);
+  }
+
+  async read(): Promise<string | undefined> {
+    const cacheFileName = this.getCacheFileName();
```
Comment (nabeelsaabna) [inline lib/util/cache/repository/impl/client/local/index.ts:19]:
```suggestion
```

**Review by nabeelsaabna — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md (hunk 4 above)
Comment (Gabriel-Ladzaretti) [reply docs/usage/self-hosted-configuration.md:629]: no, when auto generated the description from `lib/config/options/index.ts` is appended. i just dont have much to add apart from what is already written there

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.ts:19]: this was originally present before this pr, i guess this have something to do with exceptions and the error message omitted in the catch block. `cacheFileName` is used there. 
this was basically extracted as is from the original code

File: lib/util/cache/repository/impl/client/cache-client-factory.ts (hunk 6 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/cache-client-factory.ts:35]: to change clients I guess and testing. the factory is static, so we will not be able to change clients in such cases.

File: lib/config/options/index.ts (hunk 5 above)
Comment (viceice) [reply lib/config/options/index.ts]: we should set to enum like and add allowed values

File: lib/util/cache/repository/impl/client/cache-client-factory.ts (hunk 3 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/cache-client-factory.ts]: ive also run comparison between logs from branch main run vs logs from this feature branch run (same config same repo).
logs are basically the same other than timestamps that is.

thats i guess the most important log message in this context
image

File: lib/config/options/index.ts (hunk 5 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/options/index.ts]: we have additional `repositoryCacheType`, which defaults to 'local' but it is intended to be 
`'local' | 's3' | 'redis'`
if thats what you mean

Comment (Gabriel-Ladzaretti) [reply lib/config/options/index.ts]: didnt want to change existing behavior + config options.

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -229,6 +229,16 @@ const options: RenovateOptions[] = [
...
   },
+  {
+    name: 'repositoryCacheType',
+    description:
+      'Set the type of renovate repository cache if repositoryCache is not disabled.',
+    globalOnly: true,
+    type: 'string',
+    stage: 'repository',
+    default: 'local',
+    experimental: true,
```
Comment (viceice) [inline lib/config/options/index.ts:240]: add allowed values even if it's only one

File: lib/util/cache/repository/impl/client/cache-client-factory.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    const platform = GlobalConfig.get('platform');
+
+    switch (cacheType) {
+      case 'local':
+        this.client = new LocalRepoCache(platform!, repository);
+        break;
+      // istanbul ignore next: untestable
+      default:
+        this.client = new LocalRepoCache(platform!, repository);
+        logger.warn(
```
Comment (Gabriel-Ladzaretti) [inline lib/util/cache/repository/impl/client/cache-client-factory.ts:24]:
```suggestion
    const platform = GlobalConfig.get('platform')!;

    switch (cacheType) {
      case 'local':
        this.client = new LocalRepoCache(platform, repository);
        break;
      // istanbul ignore next: untestable
      default:
        this.client = new LocalRepoCache(platform, repository);
        logger.warn(
```

File: lib/config/options/index.ts (hunk 8 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/options/index.ts:240]: got it, now i understand what you meant 😅

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.ts:19]: remove inner duplicate so the cache file name is in scope in the catch block

File: lib/util/cache/repository/init.ts [hunk 10]
```diff
@@ -10,23 +10,23 @@ export async function initRepoCache(config: RenovateConfig): Promise<void> {
   resetCache();
 
   const { platform } = GlobalConfig.get();
```
Comment (rarkins) [inline lib/util/cache/repository/init.ts:12]: Looks like `platform` is no longer used in this file and its references can be removed?

**Review by rarkins — DISMISSED** [review]
Have you got any working prototype for your planned S3 cache which would allow you to validate that you've performed the appropriate refactoring/abstractions here?

File: lib/util/cache/repository/init.ts (hunk 10 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/init.ts:12]: it is still used as part of the first `if` check.
https://github.com/renovatebot/renovate/blob/2d71049650dfee6a8ba65c13b627422db79830b1/lib/util/cache/repository/init.ts#L15-L17

and I'm reliant on this check in some inner logic as i need the `platform` api

Comment (viceice) [reply lib/util/cache/repository/init.ts:12]: platform should never be undefined here

Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/init.ts:12]: so should this check be removed? its part of the existing code so does the permalink above

Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/init.ts:12]: internally im using the `not null assertion` (!)

**Comment by Gabriel-Ladzaretti** [comment]
here is a working s3 client implementation 
https://github.com/Gabriel-Ladzaretti/renovate/commit/ac100a9c1fbbe5eaafd6b427976d73c717ad20b1
connection details:

aws related credentials are passed as ENV variables:

```
AWS_ACCESS_KEY_ID
AWS_SECRET_ACCESS_KEY
AWS_SESSION_TOKEN
```

bucket and region are hardcoded and we probably need to figure out how to pass them.

```
DEBUG: initRepo("ladzaretti/alpine-node") (repository=ladzaretti/alpine-node)
DEBUG: ladzaretti/alpine-node default branch = main (repository=ladzaretti/alpine-node)
DEBUG: Using personal access token for git init (repository=ladzaretti/alpine-node)
DEBUG: S3RepoCache: read success (repository=ladzaretti/alpine-node)
DEBUG: Repository cache is valid (repository=ladzaretti/alpine-node)

```

**Comment by Gabriel-Ladzaretti** [comment]
Note that there is a key miss match between the app and the cli.
the app uses `gh/${repository}/cache.json`

while the cli uses `${this.platform}/${this.repository}/cache.json`
where in Github's case `this.platform = github`

**Comment by nabeelsaabna** [comment]
> ```
> repositoryCacheType = 'local' | string;
> 
> // for 'local'
> repositoryCacheType  = 'local';
> 
> // for s3
> repositoryCacheType  = 's3, region, bucket';
> 
> // for redis
> repositoryCacheType  = 'redis, url, etc.';
> ```

`repositoryCacheType` value should b simple (`local | s3 | redis`), but this is a discussion for another PR 😃

**Review by nabeelsaabna — DISMISSED** [review]

File: lib/util/cache/repository/impl/client/local/index.spec.ts [hunk 11]
```diff
@@ -20,25 +21,24 @@ async function createCacheRecord(
   const hash = hasha(jsonStr, { algorithm: 'sha256' });
   const compressed = await compress(jsonStr);
   const payload = compressed.toString('base64');
-  const record: RepoCacheRecord = { revision, repository, payload, hash };
-  return record;
+  return { revision, repository, payload, hash };
 }
 
-describe('util/cache/repository/impl/local', () => {
+describe('util/cache/repository/impl/client/local/index', () => {
```
Comment (viceice) [inline lib/util/cache/repository/impl/client/local/index.spec.ts:27]: why we need the client sub folder?

File: lib/util/cache/repository/impl/client/local/index.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import upath from 'upath';
+import { GlobalConfig } from '../../../../../../config/global';
+import { logger } from '../../../../../../logger';
+import { outputCacheFile, readCacheFile } from '../../../../../fs';
+import type { CacheClient, RepoCacheRecord } from '../../../types';
+
+export class LocalRepoCache implements CacheClient {
+  constructor(private platform: string, private repository: string) {}
```
Comment (viceice) [inline lib/util/cache/repository/impl/client/local/index.ts:8]: shouldn't those args passed to parent class as we need them on all implementations?

File: lib/util/cache/repository/impl/repository-cache-impl.ts [hunk 13]
```diff
@@ -1,43 +1,36 @@
...
-import type { RepoCacheRecord } from '../types';
+import type { CacheClient } from '../types';
 import { RepoCacheBase } from './base';
+import { CacheClientFactory } from './client/cache-client-factory';
 
 const compress = promisify(zlib.brotliCompress);
 const decompress = promisify(zlib.brotliDecompress);
 
-export class LocalRepoCache extends RepoCacheBase {
+export class RepositoryCacheImpl extends RepoCacheBase {
```
Comment (viceice) [inline lib/util/cache/repository/impl/repository-cache-impl.ts:19]: i think the base class should now merged into this.

previously it was planned that all implementations extend from base and don't introduce another layer of classes.

can you refactor to only have this:
- one factory to create implementation
- one base class
- one local implementation
- one null implementation (for disabled cache)
- maybe an interface like below

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 12 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.ts:8]: now sure i follow, these are are passed and are available in the factory class. so it will be available for all implementations.

for example here is the factory in my S3 client implementation
https://github.com/Gabriel-Ladzaretti/renovate/blob/4a9e925f32341b76ac45b352586e737a03f513cc/lib/util/cache/repository/impl/client/cache-client-factory.ts#L8

File: lib/util/cache/repository/impl/client/local/index.spec.ts (hunk 11 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.spec.ts:27]: to group all client implementations and differentiate them from the other factory and cache impl. 
can be removed if desired, its just a preference

**Comment by rarkins** [comment]
For future: I would like us to assume that the AWS S3 client can get credentials from the environment, either through exposed env vars or other means (e.g. IAM roles). So then for configuring S3 we can do "repositoryCacheType": "s3://bucket-name"

Another possibility is to use `repositoryCache` instead (`null` default, or `file` or `s3://...` values) and add a new option `repositoryCacheReset` as boolean

File: lib/util/cache/repository/impl/client/local/index.spec.ts (hunk 11 above)
Comment (viceice) [reply lib/util/cache/repository/impl/client/local/index.spec.ts:27]: i don't think we'll have any other types here in future, so remove.

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 12 above)
Comment (viceice) [reply lib/util/cache/repository/impl/client/local/index.ts:8]: pass them to base class and make them protected properties

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 1 above)
Comment (viceice) [reply lib/util/cache/repository/impl/client/local/index.ts:9]: move that to the cache base class and remove this class. we don't need so many layers.

in this pr we now have
- index
- RepositoryCacheImpl
- CacheClient

so i prefer that the init.ts calls the Factory and calls index.ts/setCache with result.
That way we can completle remove `RepositoryCacheImpl`, funtionallity should be moved to `RepoCacheBase` and all clients should extend that.

File: lib/util/cache/repository/impl/client/local/index.spec.ts (hunk 11 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.spec.ts:27]: are you referring to the `local` folder i.e. rename `index.ts` -> `local.ts`.
in this case we do need the `client` folder

File: lib/util/cache/repository/impl/client/local/index.ts (hunk 12 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.ts:8]: `repository` already exists on at the `RepositoryCacheImpl` class, platform is available from the global config object.
should i add the platform explicitly even so?

Comment (viceice) [reply lib/util/cache/repository/impl/client/local/index.ts:8]: platform can probably fetched in contructor of base class from global config and made available as protected property. Or just fetch it from global config when it's needed.

Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/client/local/index.ts:8]: >Or just fetch it from global config when it's needed.

Ok, that's exactly what is done currently, just the repo prop is not protected instead of being private

File: lib/util/cache/repository/impl/repository-cache-impl.ts (hunk 13 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/repository-cache-impl.ts]: * one factory - done
* one base class - `RepositoryCacheImpl` now implements `RepoCache` instead of extending `RepoCacheBase`
* one local implementation - done
* one null implementation - turned base to this
* maybe an interface like below - we have `RepoCache` & `CacheClient` did you mean anything else?

**Comment by Gabriel-Ladzaretti** [comment]
> either through exposed env vars or other means (e.g. IAM roles). So then for configuring S3 we can do "repositoryCacheType": "s3://bucket-name"

S3 client that implements the above

needed ENV vars:
```
AWS_ACCESS_KEY_ID
AWS_SECRET_ACCESS_KEY
AWS_SESSION_TOKEN
AWS_REGION
```
`repositoryCacheType` is either 'local', `s3://bucket` and probably will also be `redis://...` in the future.

https://github.com/Gabriel-Ladzaretti/renovate/commit/037a5bd4bf08dd6ae34e7c9cfc94d7738fcdc9ed

ill continue this once current PR is merged.

File: lib/config/options/index.ts [hunk 14]
```diff
@@ -229,6 +229,17 @@ const options: RenovateOptions[] = [
     default: 'disabled',
     experimental: true,
   },
+  {
+    name: 'repositoryCacheType',
+    description:
+      'Set the type of renovate repository cache if repositoryCache is not disabled.',
+    globalOnly: true,
+    type: 'string',
+    allowedValues: ['local'],
```
Comment (viceice) [inline lib/config/options/index.ts:238]: this won't work with `s3://...` values later

File: lib/config/types.ts [hunk 15]
```diff
@@ -11,6 +11,7 @@ export type RenovateConfigStage =
   | 'pr';
 
 export type RepositoryCacheConfig = 'disabled' | 'enabled' | 'reset';
+export type RepositoryCacheType = 'local'; // add "s3" and "redis" once implemented
```
Comment (viceice) [inline lib/config/types.ts:14]: same

File: lib/util/cache/repository/impl/cache-client-factory.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
+import { GlobalConfig } from '../../../../config/global';
+import type { RepositoryCacheType } from '../../../../config/types';
+import { logger } from '../../../../logger';
+import type { CacheClient } from '../types';
+import { LocalRepoCache } from './client/local';
+
+export class CacheClientFactory {
```
Comment (viceice) [inline lib/util/cache/repository/impl/cache-client-factory.ts:7]:
```suggestion
export class CacheFactory {
```

remove `client` from all other too

File: lib/util/cache/repository/impl/cache-client-factory.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  private static client: CacheClient | null;
+
+  static get(repository: string, cacheType: RepositoryCacheType): CacheClient {
+    if (this.client) {
+      return this.client;
+    }
```
Comment (viceice) [inline lib/util/cache/repository/impl/cache-client-factory.ts:13]: don't cache here

File: lib/util/cache/repository/impl/cache-client-factory.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+    const platform = GlobalConfig.get('platform')!;
+
+    switch (cacheType) {
+      case 'local':
+        this.client = new LocalRepoCache(platform, repository);
+        break;
+      // istanbul ignore next: untestable
```
Comment (viceice) [inline lib/util/cache/repository/impl/cache-client-factory.ts:21]: it's testable, just call with `undefined as RepositoryCacheType` in test

File: lib/util/cache/repository/impl/repository-cache-impl.ts [hunk 19]
```diff
@@ -1,43 +1,35 @@
...
-import type { RepoCacheRecord } from '../types';
-import { RepoCacheBase } from './base';
+import type { CacheClient, RepoCache, RepoCacheData } from '../types';
+import { CacheClientFactory } from './cache-client-factory';
 
 const compress = promisify(zlib.brotliCompress);
 const decompress = promisify(zlib.brotliDecompress);
 
-export class LocalRepoCache extends RepoCacheBase {
+export class RepositoryCacheImpl implements RepoCache {
```
Comment (viceice) [inline lib/util/cache/repository/impl/repository-cache-impl.ts:18]:
```suggestion
export class RepositoryCacheBase implements RepoCache {
```

this should be the base class for all cache implementations (beside null cache), it should have 2 abstract protected methods for `_load` and `_save` which is called from this class

File: lib/util/cache/repository/impl/repository-cache-null-impl.ts [hunk 20]
```diff
@@ -1,6 +1,6 @@
 import type { RepoCache, RepoCacheData } from '../types';
 
-export class RepoCacheBase implements RepoCache {
+export class RepositoryCacheNullImpl implements RepoCache {
```
Comment (viceice) [inline lib/util/cache/repository/impl/repository-cache-null-impl.ts:3]:
```suggestion
export class NullRepositoryCache implements RepoCache {
```

remove `Impl` suffix from all

File: lib/util/cache/repository/init.ts [hunk 21]
```diff
@@ -9,24 +9,24 @@ import { resetCache, setCache } from '.';
...
+  if (repositoryCache === 'disabled') {
+    setCache(new RepositoryCacheNullImpl());
     return;
   }
 
   if (repositoryCache === 'enabled') {
-    const localCache = new LocalRepoCache(platform, repository);
-    await localCache.load();
-    setCache(localCache);
+    const cache = new RepositoryCacheImpl(repository!, type);
```
Comment (viceice) [inline lib/util/cache/repository/init.ts:20]:
```suggestion
    const cache = CacheFactory.get(repository!, type);
```

File: lib/util/cache/repository/init.ts [hunk 22]
```diff
@@ -9,24 +9,24 @@ import { resetCache, setCache } from '.';
...
+    await cache.load();
+    setCache(cache);
     return;
   }
 
   if (repositoryCache === 'reset') {
-    const localCache = new LocalRepoCache(platform, repository);
-    await localCache.save();
-    setCache(localCache);
+    const cache = new RepositoryCacheImpl(repository!, type);
```
Comment (viceice) [inline lib/util/cache/repository/init.ts:27]:
```suggestion
    const cache = CacheFactory.get(repository!, type);
```

**Review by viceice — COMMENTED** [review]
hopefully it's more clear now, ortherwise i should do the refactoring myself

File: lib/config/options/index.ts (hunk 14 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/options/index.ts:238]: i know https://github.com/Gabriel-Ladzaretti/renovate/commit/037a5bd4bf08dd6ae34e7c9cfc94d7738fcdc9ed

File: lib/config/types.ts (hunk 15 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/types.ts:14]: same

File: lib/config/options/index.ts (hunk 14 above)
Comment (Gabriel-Ladzaretti) [reply lib/config/options/index.ts:238]: we can discuss the s3 and redis whenever we are implementing it

Comment (viceice) [reply lib/config/options/index.ts:238]: lets remove it here, when it's removed in next pr anyways

File: lib/config/types.ts (hunk 15 above)
Comment (viceice) [reply lib/config/types.ts:14]: lets add string now, even it it's no used yet, makes testing not supported implementations easier

File: lib/util/cache/repository/impl/repository-cache-impl.ts (hunk 19 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/repository-cache-impl.ts:18]: @viceice 
This is what I intend to implement, hope this is what you meant
image

_load and _save through me off quite a bit as these are cache methods, while the clients have read/write

Comment (viceice) [reply lib/util/cache/repository/impl/repository-cache-impl.ts:18]: sounds good, function name was a sample, read and write is better. 👍

File: lib/util/cache/repository/impl/cache-factory.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+export class CacheFactory {
+  private static client: RepoCache | null;
+
+  static get(
+    repository: string,
+    cacheType: RepositoryCacheType = 'local'
+  ): RepoCache {
+    const type = parseUrl(cacheType)?.protocol ?? 'local';
+
```
Comment (Gabriel-Ladzaretti) [inline lib/util/cache/repository/impl/cache-factory.ts:15]:
```suggestion
    const type = cacheType.split('://')[0].trim().toLowerCase();
```
couldn't choose between these two

File: lib/util/cache/repository/impl/local.ts [hunk 24]
```diff
@@ -1,27 +1,13 @@
...
 
-const compress = promisify(zlib.brotliCompress);
-const decompress = promisify(zlib.brotliDecompress);
-
-export class LocalRepoCache extends RepoCacheBase {
-  private oldHash: string | null = null;
-
-  constructor(private platform: string, private repository: string) {
-    super();
+export class LocalRepositoryCache extends RepositoryCacheBase {
```
Comment (nabeelsaabna) [inline lib/util/cache/repository/impl/local.ts:8]: the word "repository" a bit annoying in this case
maybe 

- RepositoryCacheLocal
- RepositoryLocalCache

Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/local.ts:8]: agree, renamed all to use the original naming scheme files included.

File: lib/util/cache/repository/impl/cache-factory.ts (hunk 23 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/cache-factory.ts]: removed for now

File: lib/util/cache/repository/impl/cache-factory.ts [hunk 25]
```diff
@@ -0,0 +1,28 @@
+import type { RepositoryCacheType } from '../../../../config/types';
+import { logger } from '../../../../logger';
+import type { RepoCache } from '../types';
+import { RepoCacheLocal } from './local';
+
+export class CacheFactory {
+  private static client: RepoCache | null;
```
Comment (viceice) [inline lib/util/cache/repository/impl/cache-factory.ts:7]: don't cache here, should always return a fresh value
```suggestion
```

File: lib/util/cache/repository/impl/cache-factory.ts [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+      default:
+        this.client = new RepoCacheLocal(repository);
+        logger.warn(
+          { cacheType },
+          `Repository cache type not supported using type "local" instead`
+        );
+        break;
+    }
+
+    return this.client;
```
Comment (viceice) [inline lib/util/cache/repository/impl/cache-factory.ts:26]:
```suggestion
    switch (cacheType) {
      case 'local':
        return new RepoCacheLocal(repository);
      default:        
        logger.warn(
          { cacheType },
          `Repository cache type not supported using type "local" instead`
        );
        return new RepoCacheLocal(repository);
    }
```

doesn't make sense to cache, otherwise we always get the same instance for different repositories!

File: lib/util/cache/repository/impl/null.ts [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+
+  // istanbul ignore next
+  async load(): Promise<void> {
+    await Promise.resolve();
+  }
+
+  // istanbul ignore next
+  async save(): Promise<void> {
+    await Promise.resolve();
+  }
```
Comment (viceice) [inline lib/util/cache/repository/impl/null.ts:14]:
```suggestion
  load(): Promise<void> {
    return Promise.resolve();
  }

  // istanbul ignore next
  save(): Promise<void> {
    return Promise.resolve();
  }
```

File: lib/util/cache/repository/impl/base.ts [hunk 28]
```diff
@@ -1,16 +1,74 @@
...
+  private data: RepoCacheData = {};
+
+  protected constructor(protected readonly repository: string) {}
+
+  protected abstract read(): Promise<string | undefined>;
+
+  protected abstract write(data: RepoCacheRecord): Promise<void>;
 
-  // istanbul ignore next
   async load(): Promise<void> {
```
Comment (zharinov) [inline lib/util/cache/repository/impl/base.ts:28]: I think my naming wasn't good, maybe it's a good chance to rename methods to something like `load/save` and `encode/decode` (or `compress/decompress`)

Comment (zharinov) [reply lib/util/cache/repository/impl/base.ts:28]: Or maybe, `loadRaw/saveRaw` for the abstract methods above

File: lib/config/types.ts [hunk 29]
```diff
@@ -11,6 +11,7 @@ export type RenovateConfigStage =
   | 'pr';
 
 export type RepositoryCacheConfig = 'disabled' | 'enabled' | 'reset';
+export type RepositoryCacheType = 'local' | string;
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/config/types.ts:14]: https://github.com/renovatebot/renovate/pull/17146#issuecomment-1214690058
we want it to support s3/redis urls in the future e.g `s3://bucket-name`. i think it is easier than changing existing `repositoryCache` flag

File: lib/util/cache/repository/impl/base.ts (hunk 28 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/cache/repository/impl/base.ts:28]: I do like the `save/load` for cache methods, and `read/write` for the actual reading/writing to fs/storage.

File: lib/config/types.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/config/types.ts:14]: but then again, we can remove it now and discuss it in the s3/redis implementation PR ¯\\_(ツ)_/¯

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/config/types.ts:14]: Hm, URL was unexpected for me to see with this type. In this context, I wonder if we may want to also support `file://...` scheme then.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/config/types.ts:14]: leave as is for now

File: lib/util/cache/repository/impl/base.ts (hunk 28 above)
Comment (viceice) [reply lib/util/cache/repository/impl/base.ts:28]: me too

Comment (zharinov) [reply lib/util/cache/repository/impl/base.ts:28]: Well, okay

**Review by viceice — APPROVED** [review]

**Review by nabeelsaabna — APPROVED** [review]
